In [2]:
import pandas as pd
import numpy as np
import sys
from pytrends.request import TrendReq
import time
import warnings
warnings.filterwarnings("ignore")
# sys.path.append('../methods')
# from SVI_methods import index_with_dates

from datetime import datetime

C:\Users\x4287225\AppData\Local\Temp\ipykernel_11024\3802138766.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [5]:


def index_with_dates(keywords, start_date, end_date):
     """
     keywords: list of keywords => ['AAPL']
     start_date: a date as type string => '2018-01-01'
     end_date: a date as type string => '2018-01-01'
     when time interval is under than 8 months it returns on a daily basis
     when time interval is under than 6 years it returns on a weekly basis
     when time interval more than 6 years it returns on a yearly basis
     # cat='' Financial Markets: 1163
     """
     try:
          pytrends = TrendReq(hl='en-US', tz=360, requests_args={'verify':False})
          timeframe = start_date + " " + end_date
          pytrends.build_payload(kw_list=keywords, timeframe=timeframe, geo='', gprop='')
          results = pytrends.interest_over_time()
          return results
     except Exception as e:
          print(e)
          print(f'error has occured for {keywords}, {start_date} to {end_date}')

### Monthly SVI Extraction

In [6]:
companies = pd.read_csv("../data/companies_after_2005.csv")
symbols = np.array(companies['Symbol'])

In [28]:
# monthly_svi = pd.DataFrame(columns = ['symbol','date', 'svi'])

bad_symbols = ['ADBE', 'ACN', 'PEP']
empty_symbols = []
for s in symbols[87:]: 
     time.sleep(6)
     x = index_with_dates([s], "2005-01-01", "2023-12-01")
     if isinstance(x, pd.DataFrame):
          if(len(x) != 0):
               new_df = pd.DataFrame(columns = ['symbol','date', 'svi'])
               new_df['date'] = x.index
               new_df['svi'] = x.iloc[:,0].values
               new_df['symbol'] = s
               monthly_svi = pd.concat([monthly_svi,new_df], axis = 0)
               monthly_svi.reset_index(inplace=True, drop=True)
          else:
               print(s, "is an empty symbol")
               empty_symbols.append(s)
     else:
          print(s, "has an error")
          bad_symbols.append(s)


6861.T is an empty symbol
8035.T is an empty symbol
601628.SS is an empty symbol
AI.PA is an empty symbol
601318.SS is an empty symbol
The request failed: Google returned a response with code 429
error has occured for ['ADI'], 2005-01-01 to 2023-12-01
ADI has an error
ABBN.SW is an empty symbol
4063.T is an empty symbol
SBIN.NS is an empty symbol
BHARTIARTL.NS is an empty symbol
ATCO-B.ST is an empty symbol
000660.KS is an empty symbol
6501.T is an empty symbol
INVE-B.ST is an empty symbol
ZURN.SW is an empty symbol
The request failed: Google returned a response with code 429
error has occured for ['MRK.DE'], 2005-01-01 to 2023-12-01
MRK.DE has an error
The request failed: Google returned a response with code 429
error has occured for ['IBE.MC'], 2005-01-01 to 2023-12-01
IBE.MC has an error
002594.SZ is an empty symbol
The request failed: Google returned a response with code 429
error has occured for ['PH'], 2005-01-01 to 2023-12-01
PH has an error
The request failed: Google returned a

In [32]:
very_bad_symbols = []
# empty_symbols = []
for s in bad_symbols: 
     time.sleep(20)
     x = index_with_dates([s], "2005-01-01", "2023-12-01")
     if isinstance(x, pd.DataFrame):
          if(len(x) != 0):
               new_df = pd.DataFrame(columns = ['symbol','date', 'svi'])
               new_df['date'] = x.index
               new_df['svi'] = x.iloc[:,0].values
               new_df['symbol'] = s
               monthly_svi = pd.concat([monthly_svi,new_df], axis = 0)
               monthly_svi.reset_index(inplace=True, drop=True)
          else:
               print(s, "is an empty symbol")
               empty_symbols.append(s)
     else:
          print(s, "has an error")
          very_bad_symbols.append(s)

MRK.DE is an empty symbol
IBE.MC is an empty symbol
601899.SS is an empty symbol
NDA-FI.HE is an empty symbol
DB1.DE is an empty symbol
600276.SS is an empty symbol
601998.SS is an empty symbol


In [34]:
monthly_svi['symbol'].nunique()

392

In [36]:
monthly_svi.to_csv("../data/monthly_svi.csv")
np.save("../data/empty_symbols.npy", empty_symbols)

### Weekly SVI

* We have 19 years, the divide will be 2005/1 to 2009/12, 2010/1 to 2014/12, 2015/1 to 2019/12, 2020/1 to 2023/12 . Th reasons is because of the fact that we need weekly data and the time intervals should be less than 6 years.
* We have 392 symbols


In [61]:
monthly_svi['symbol'].unique()

array(['MSFT', 'AAPL', 'NVDA', 'AMZN', 'GOOG', 'BRK-B', 'LLY', 'TSM',
       'NVO', 'JPM', 'UNH', 'WMT', 'XOM', 'JNJ', 'PG', 'ASML', 'HD',
       'TCEHY', 'MRK', 'COST', 'TM', 'ORCL', 'CVX', 'AMD', 'CRM', 'BAC',
       'KO', 'NFLX', 'TMO', 'MCD', 'SAP', 'DIS', '601857.SS', 'AZN',
       'ABT', 'CSCO', 'TMUS', 'WFC', 'DHR', 'INTC', 'INTU', 'IBM', 'QCOM',
       'VZ', 'AMAT', 'CMCSA', 'CAT', 'NKE', 'PFE', 'AXP', 'AMGN', 'UNP',
       'TXN', 'SIE.DE', 'MS', 'RY', 'HDB', 'ISRG', 'SPGI', 'SYK', 'IDEXY',
       'LOW', 'COP', 'HON', 'BKNG', 'UL', 'UPS', 'PLD', 'AIR.PA', 'GS',
       'BA', 'BUD', 'LRCX', 'T', 'RTX', 'DTE.DE', 'BLK', 'SCHW', 'NEE',
       'SNY', 'PBR', 'MDT', 'RIO', 'TJX', 'PGR', 'SONY', 'VRTX', 'TD',
       'SBUX', 'C', 'ALV.DE', 'REGN', 'ADP', 'LMT', 'DE', 'BMY', 'BP',
       'CI', 'MMC', 'CVS', 'MDLZ', 'BSX', 'EL.PA', 'CSL.AX', 'KLAC',
       'GILD', 'FI', 'MU', 'AMT', 'IBN', '9983.T', 'GSK', 'INFY', 'SNPS',
       '8058.T', 'CNI', '0883.HK', 'HCA', 'DEO', '9984.T', 'WM', 'R

In [74]:
start_intervals = ["2005-01-01", "2010-01-01", "2015-01-01", "2020-01-01"]
end_intervals = ["2009-12-31", "2014-12-31", "2019-12-31", "2023-12-31"]
weekly_symbols = monthly_svi['symbol'].unique()
event_log = pd.DataFrame(columns=['symbol','start_date', 'end_date'])
for s in weekly_symbols:
     for d in range(4):
          event = [s, start_intervals[d],end_intervals[d]]
          event_log.loc[len(event_log)] = event
          

event_log


symbol  start_date    end_date
0      MSFT  2005-01-01  2009-12-31
1      MSFT  2010-01-01  2014-12-31
2      MSFT  2015-01-01  2019-12-31
3      MSFT  2020-01-01  2023-12-31
4      AAPL  2005-01-01  2009-12-31
...     ...         ...         ...
1563    HUM  2020-01-01  2023-12-31
1564   GIII  2005-01-01  2009-12-31
1565   GIII  2010-01-01  2014-12-31
1566   GIII  2015-01-01  2019-12-31
1567   GIII  2020-01-01  2023-12-31

[1568 rows x 3 columns]

In [75]:
Weekly_svi = pd.DataFrame(columns = ['symbol','date', 'svi'])
error_log = pd.DataFrame(columns=['symbol','start_date', 'end_date', 'error'])

for r in range(len(event_log)):
     time.sleep(6)
     x = index_with_dates([event_log['symbol'][r]], event_log['start_date'][r], event_log['end_date'][r])
     if isinstance(x, pd.DataFrame):
          if(len(x) != 0):
               new_df = pd.DataFrame(columns = ['symbol','date', 'svi'])
               new_df['date'] = x.index
               new_df['svi'] = x.iloc[:,0].values
               new_df['symbol'] = event_log['symbol'][r]
               Weekly_svi = pd.concat([Weekly_svi,new_df], axis = 0)
               Weekly_svi.reset_index(inplace=True, drop=True)
          else:
               print(event_log['symbol'][r], "is an empty symbol")
               error_log.loc[len(error_log)] = [event_log['symbol'][r], event_log['start_date'][r], event_log['end_date'][r], "empty symbol"]
     else:
          print(event_log['symbol'][r], "has an error")
          error_log.loc[len(error_log)] = [event_log['symbol'][r], event_log['start_date'][r], event_log['end_date'][r], "server error"]

The request failed: Google returned a response with code 429
error has occured for ['AXP'], 2015-01-01 to 2019-12-31
AXP has an error
The request failed: Google returned a response with code 429
error has occured for ['REGN'], 2020-01-01 to 2023-12-31
REGN has an error
The request failed: Google returned a response with code 429
error has occured for ['BMY'], 2015-01-01 to 2019-12-31
BMY has an error
The request failed: Google returned a response with code 429
error has occured for ['MRVL'], 2015-01-01 to 2019-12-31
MRVL has an error
The request failed: Google returned a response with code 429
error has occured for ['HES'], 2005-01-01 to 2009-12-31
HES has an error
The request failed: Google returned a response with code 429
error has occured for ['BAS.DE'], 2015-01-01 to 2019-12-31
BAS.DE has an error
The request failed: Google returned a response with code 429
error has occured for ['STM'], 2020-01-01 to 2023-12-31
STM has an error


In [78]:
Weekly_svi['symbol'].nunique()

392

In [81]:
Weekly_svi

symbol       date svi
0        MSFT 2005-01-02  18
1        MSFT 2005-01-09  19
2        MSFT 2005-01-16  12
3        MSFT 2005-01-23  20
4        MSFT 2005-01-30  15
...       ...        ...  ..
387136   GIII 2023-12-03  71
387137   GIII 2023-12-10  52
387138   GIII 2023-12-17  47
387139   GIII 2023-12-24  43
387140   GIII 2023-12-31  64

[387141 rows x 3 columns]

In [80]:
error_log

symbol  start_date    end_date         error
0     AXP  2015-01-01  2019-12-31  server error
1    REGN  2020-01-01  2023-12-31  server error
2     BMY  2015-01-01  2019-12-31  server error
3    MRVL  2015-01-01  2019-12-31  server error
4     HES  2005-01-01  2009-12-31  server error
5  BAS.DE  2015-01-01  2019-12-31  server error
6     STM  2020-01-01  2023-12-31  server error

In [82]:
# Weekly_svi = pd.DataFrame(columns = ['symbol','date', 'svi'])
error_log_second = pd.DataFrame(columns=['symbol','start_date', 'end_date', 'error'])

for r in range(len(error_log)):
     time.sleep(30)
     x = index_with_dates([error_log['symbol'][r]], error_log['start_date'][r], error_log['end_date'][r])
     if isinstance(x, pd.DataFrame):
          if(len(x) != 0):
               new_df = pd.DataFrame(columns = ['symbol','date', 'svi'])
               new_df['date'] = x.index
               new_df['svi'] = x.iloc[:,0].values
               new_df['symbol'] = error_log['symbol'][r]
               Weekly_svi = pd.concat([Weekly_svi,new_df], axis = 0)
               Weekly_svi.reset_index(inplace=True, drop=True)
          else:
               print(error_log['symbol'][r], "is an empty symbol")
               error_log_second.loc[len(error_log_second)] = [error_log['symbol'][r], error_log['start_date'][r], error_log['end_date'][r], "empty symbol"]
     else:
          print(error_log['symbol'][r], "has an error")
          error_log_second.loc[len(error_log_second)] = [error_log['symbol'][r], error_log['start_date'][r], error_log['end_date'][r], "server error"]

In [83]:
Weekly_svi

symbol       date svi
0        MSFT 2005-01-02  18
1        MSFT 2005-01-09  19
2        MSFT 2005-01-16  12
3        MSFT 2005-01-23  20
4        MSFT 2005-01-30  15
...       ...        ...  ..
388859    STM 2023-12-03   0
388860    STM 2023-12-10   0
388861    STM 2023-12-17   0
388862    STM 2023-12-24   0
388863    STM 2023-12-31   0

[388864 rows x 3 columns]

In [84]:
Weekly_svi.to_csv("../data/weekly_svi.csv")

### after getting the weekly and monthly data, we need to normalize 

In [3]:
weekly_svi = pd.read_csv("../data/weekly_svi.csv")
monthly_svi = pd.read_csv("../data/monthly_svi.csv")

weekly_svi.head(3)
symbols = weekly_svi['symbol'].unique()

### 2005/1 to 2009/12, 2010/1 to 2014/12, 2015/1 to 2019/12, 2020/1 to 2023/12

In [8]:
# '2005-01' in
monthly_svi[monthly_svi['date'] == '2005-01-01'].index

Index([    0,   228,   456,   684,   912,  1140,  1368,  1596,  1824,  2052,
       ...
       87096, 87324, 87552, 87780, 88008, 88236, 88464, 88692, 88920, 89148],
      dtype='int64', length=392)

In [4]:
# first_cat = 2005/1 to 2009/12
# second_cat = 2010/1 to 2014/12
# thirs_cat = 2015/1 to 2019/12
# forth_cat = 2020/1 to 2023/12
def categorize (date_str):
     # Converting the string to datetime object
     date_object = datetime.strptime(date_str, '%Y-%m-%d')
     # Extracting the year from the datetime object
     year_object = date_object.year

     # Comparison
     if year_object < 2010 :
          return 1
     elif year_object < 2015 :
          return 2
     elif year_object < 2020 :
          return 3
     else :
          return 4
     

In [5]:
monthly_svi['date_category'] = monthly_svi['date'].apply(lambda x: categorize(x))
weekly_svi['date_category'] = weekly_svi['date'].apply(lambda x: categorize(x))

In [16]:
sum(monthly_svi[monthly_svi['symbol'] == 'MSFT'][monthly_svi['date_category'] == 1]['svi'])

1741

In [24]:
monthly_svi

Unnamed: 0 symbol        date  svi  date_category
0               0   MSFT  2005-01-01   12              1
1               1   MSFT  2005-02-01   12              1
2               2   MSFT  2005-03-01   12              1
3               3   MSFT  2005-04-01   28              1
4               4   MSFT  2005-05-01   25              1
...           ...    ...         ...  ...            ...
89371       89371   GIII  2023-08-01    0              4
89372       89372   GIII  2023-09-01    0              4
89373       89373   GIII  2023-10-01    0              4
89374       89374   GIII  2023-11-01    0              4
89375       89375   GIII  2023-12-01    0              4

[89376 rows x 5 columns]

In [30]:
category_score = pd.DataFrame(columns = ['symbol', 'cate_1', 'cate_2', 'cate_3', 'cate_4'])
category_score['symbol'] = symbols

for s in symbols:
     monthly = monthly_svi[monthly_svi['symbol'] == s]
     category_score.loc[category_score['symbol'] == s, 'cate_1'] = sum(monthly[monthly['date_category'] == 1]['svi'])
     category_score.loc[category_score['symbol'] == s, 'cate_2'] = sum(monthly[monthly['date_category'] == 2]['svi'])
     category_score.loc[category_score['symbol'] == s, 'cate_3'] = sum(monthly[monthly['date_category'] == 3]['svi'])
     category_score.loc[category_score['symbol'] == s, 'cate_4'] = sum(monthly[monthly['date_category'] == 4]['svi'])
     
category_score

symbol cate_1 cate_2 cate_3 cate_4
0     MSFT   1741    784   1272   2475
1     AAPL   2206   2594   1966   1706
2     NVDA    314    244    976   1817
3     AMZN    282    397   2216   2191
4     GOOG   2167   4496   2420   1058
..     ...    ...    ...    ...    ...
387   PYPL     36      9    513   1700
388    VLO   2972   1319    844      0
389   BA.L      0   1515   2658   1430
390    HUM   2847   1403      0      0
391   GIII      0   3182   1846      0

[392 rows x 5 columns]

In [45]:
def score_cal(symbol, cat):
     if cat == 1:
          return category_score.loc[category_score['symbol'] == symbol, 'cate_1'].iloc[0]
     elif cat == 2:
          return category_score.loc[category_score['symbol'] == symbol, 'cate_2'].iloc[0]
     elif cat == 3:
          return category_score.loc[category_score['symbol'] == symbol, 'cate_3'].iloc[0]
     else:
          return category_score.loc[category_score['symbol'] == symbol, 'cate_4'].iloc[0]
     
weekly_svi['monthly_score'] = weekly_svi.apply(lambda row: score_cal(row['symbol'], row['date_category']), axis=1)

3182


In [54]:
weekly_svi['agregated_score'] = weekly_svi['svi'] * weekly_svi['monthly_score']
weekly_svi.head()

Unnamed: 0 symbol        date  svi  date_category  monthly_score  \
0           0   MSFT  2005-01-02   18              1           1741   
1           1   MSFT  2005-01-09   19              1           1741   
2           2   MSFT  2005-01-16   12              1           1741   
3           3   MSFT  2005-01-23   20              1           1741   
4           4   MSFT  2005-01-30   15              1           1741   

   agregated_score  
0            31338  
1            33079  
2            20892  
3            34820  
4            26115

In [55]:
weekly_svi.to_csv("../data/agregated_svi.csv")